### PR TITLE
More precise typing for `BuildEnvironment.domains`

### PR DESCRIPTION
**Pull request by AA-Turner**
@danieleades if you've time for review on this I'd appreciate it--originally I was using imports from `typing` rather than `typing_extensions`, but they refused to work on my computer. I'm not sure if it is my installation that's broken, or if this is a weird quirk.

### Feature or Bugfix
- Refactoring

A

### CODE REVIEW

File: sphinx/environment/__init__.py [hunk 1]
```diff
@@ -79,6 +79,60 @@
...
+        @overload
+        def __getitem__(self, key: str) -> Domain: ...  # NoQA: E704
+        def __getitem__(self, key): raise NotImplementedError  # NoQA: E704
+        def __setitem__(self, key, value): raise NotImplementedError  # NoQA: E704
+        def __delitem__(self, key): raise NotImplementedError  # NoQA: E704
+        def __iter__(self): raise NotImplementedError  # NoQA: E704
+        def __len__(self): raise NotImplementedError  # NoQA: E704
+
+else:
+    DomainsType = dict
```
Comment (danieleades) [inline sphinx/environment/__init__.py:134]: type annotations? what type of dict is this?

File: sphinx/environment/__init__.py [hunk 2]
```diff
@@ -79,6 +79,60 @@
...
+        def __getitem__(self, key: Literal["duration"]) -> DurationDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: Literal["todo"]) -> TodoDomain: ...  # NoQA: E704
+        @overload
+        def __getitem__(self, key: str) -> Domain: ...  # NoQA: E704
+        def __getitem__(self, key): raise NotImplementedError  # NoQA: E704
+        def __setitem__(self, key, value): raise NotImplementedError  # NoQA: E704
+        def __delitem__(self, key): raise NotImplementedError  # NoQA: E704
+        def __iter__(self): raise NotImplementedError  # NoQA: E704
+        def __len__(self): raise NotImplementedError  # NoQA: E704
```
Comment (danieleades) [inline sphinx/environment/__init__.py:131]: this approach doesn't seem right to me. What are you trying to achieve? you might be looking for `TypedDict`, rather than `MutableMapping`?

something like-

```python
class DomainsType(TypedDict):
    c: CDomain
    cpp: CPPDomain
    changeset: ChangeSetDomain
    ...
```

can you give me some more context here?

**Review by danieleades — CHANGES_REQUESTED** [review]

**Comment by danieleades** [comment]
> originally I was using imports from `typing` rather than `typing_extensions`, but they refused to work on my computer. I'm not sure if it is my installation that's broken, or if this is a weird quirk.

what interpreter version are you using locally?

python typing features are generally added to `typing-extensions` *first* and then stabilised into `typing`, so depending on the features you're using and the interpreter version, you might find that the features are available in typing-extensions but not in typing. The differences between build environments are likely down to different interpreter versions.

**Comment by AA-Turner** [comment]
`env.domains` holds references to all loaded domain objects, but each domain has different properties, attributes, etc. I want mypy to recognise that if I type `env.domains['std'].labels` for example that I am accessing the legal `labels` property on the `StandardDomain` object rather than the missing property on the `Domain` type. 

A `TypedDict` won't work as I understand it as users can register their own domains (see `recipie` in the docs or `duration` in `sphinx.ext.duation`)--for these cases I still want the type to be `str->Domain`.

In short--a mapping type of `str->Domain`, with certain magic keys recognised as returning subclasses of `Domain`.

A

File: sphinx/environment/__init__.py (hunk 1 above)
Comment (AA-Turner) [reply sphinx/environment/__init__.py:134]: This is the `TYPE_CHECKING is False` path, i.e. at runtime--we want to instantiate a normal dict. `mypy` was complaining if I wrote `self.domains = {}` rather than `self.domains = DomainsType()` below.

**Comment by AA-Turner** [comment]
> what interpreter version are you using locally?
> 
> python typing features are generally added to typing-extensions first and then stabilised into typing

Yep, I've contributed (albeit minorly) to TE. Running Python 3.10.6 from conda-forge, mypy 0.971. As I understand it, `mypy` vendors its own copy of typeshed, which is why I thought it might be some glitch as opposed to loading TE dynamically, but `mypy` may have changed its architecture.

A

**Comment by danieleades** [comment]
Are the 'magic keys' always set?

**Comment by AA-Turner** [comment]
From the above, all are always set apart from `duration` and `todo`. They're initialised in this loop:

https://github.com/AA-Turner/sphinx/blob/c215f522e955304ebf50829b329e7d4712eeb91f/sphinx/environment/__init__.py#L269-L271

A

File: sphinx/environment/__init__.py (hunk 1 above)
Comment (danieleades) [reply sphinx/environment/__init__.py:134]: > This is the `TYPE_CHECKING is False` path, i.e. at runtime--we want to instantiate a normal dict. `mypy` was complaining if I wrote `self.domains = {}` rather than `self.domains = DomainsType()` below.

what i mean is you haven't fully annotated the type. `dict` (or `Dict`) is equivalent to `dict[Any, Any]`, which is not what you want. Unfortunately the type annotations are used this way throughout the codebase...

can you do something like

```python
DomainsType = dict[str, Type[Domain]]
```

?

Comment (danieleades) [reply sphinx/environment/__init__.py:134]: oh i see what you mean about the *runtime* path

File: sphinx/environment/__init__.py (hunk 2 above)
Comment (danieleades) [reply sphinx/environment/__init__.py:131]: on review i'm not sure there is a better solution than what you're proposing right now. There are feature requests to allow additional keys on `TypedDict`, but there are soundness issues with the proposal.

I guess that what this really highlights is that if this codebase had used strict type annotations from the start, it would probably not be implemented this way. I presume you're avoiding breaking changes here. If you *could* make breaking changes you would just promote the magic keys to concrete properties

Comment (AA-Turner) [reply sphinx/environment/__init__.py:131]: > on review i'm not sure there is a better solution than what you're proposing right now. There are feature requests to allow additional keys on `TypedDict`, but there are soundness issues with the proposal.

Thank you for your review, this is what I feared having tried a `Protocol` and `TypedDict` approach and having gotten nowhere.

> I guess that what this really highlights is that if this codebase had used strict type annotations from the start, it would probably not be implemented this way. 

Challenging given the first commit here was around seven years before [PEP 484](https://peps.python.org/pep-0484/)!

>I presume you're avoiding breaking changes here. If you _could_ make breaking changes you would just promote the magic keys to concrete properties

I'll consider the attributes approach for Sphinx 6.0/7.0, though it has drawbacks in that the user-registered domains will be second-class in some way.

A

Comment (danieleades) [reply sphinx/environment/__init__.py:131]: > the user-registered domains will be second-class in some way.

they *are* second-class, in that they are dynamically registered, instead of statically registered. It makes sense that these would be handled differently, as this is a pretty fundamental difference. I would probably go further and make this explicit-

```python
class DomainsType:
    c: CDomain
    cpp: CPPDomain
    changeset: ChangeSetDomain
    ...
    custom: dict[str, Domain]
```

i think the only way to avoid this difference would be to somehow require users to subclass `DomainsType` to add their own domains
